src/features/bookOffer/components/BookHourChoice.tsx

```diff
@@ -39,38 +41,95 @@ export const BookHourChoice = ({ enablePricesByCategories }: Props) => {
     }
   }
 
+  const selectHour = (hour: string) => {
+    dispatch({ type: 'SELECT_HOUR', payload: hour })
+    if (!isDuo) {
+      dispatch({ type: 'SELECT_QUANTITY', payload: 1 })
+    }
+  }
+
   const changeHour = () => {
     dispatch({ type: 'CHANGE_STEP', payload: Step.HOUR })
   }
 
   const filteredStocks = useMemo(
-    () =>
-      stocks
-        .filter(({ beginningDatetime }) => {
-          if (beginningDatetime === undefined || beginningDatetime === null) return false
-          return selectedDate && beginningDatetime
-            ? formatToKeyDate(beginningDatetime) === selectedDate
-            : false
+    () => {
+      if (hasPricesStep) {
+        const distinctHours: string[] = [
+          ...new Set(
+            stocks
+              .filter(
+                (stock) =>
+                  !stock.isExpired &&
+                  stock.beginningDatetime &&
+                  formatToKeyDate(stock.beginningDatetime) === selectedDate
+              )
+              .map((stock) => stock.beginningDatetime)
+              .sort(
+                (a, b) => new Date(a as string).getTime() - new Date(b as string).getTime()
+              ) as string[]
+          ),
```

Comment:
Cette partie est un peu complexe.
Tu pourrais avoir un
```suggestion
          const stockAtDateFilterPredicate = (stock: ??) =>
            !stock.isExpired &&
            stock.beginningDatetime &&
            formatToKeyDate(stock.beginningDatetime) === selectedDate

          const sortByDatePredicate = (a: string, b: string) => {
            return new Date(a as string).getTime() - new Date(b as string).getTime()
          }

          ...new Set(
            stocks
              .filter(stockAtDateFilterPredicate)
              .map((stock) => stock.beginningDatetime)
              .sort(sortByDatePredicate) as string[]
          ),
```

Dans l'idéal il faudrait sortir les `predicate` du composant aussi.

Autre point. J'imagine qu'il y a une raison aux `as string`, mais y'a pas moyen de les corriger ?

Reply:
J'ai fait une partie de ce que tu as demandé